### PR TITLE
feat: refactor middleware stack

### PR DESCRIPTION
**Pull request by AllanFly120**
*Issue #, if available:*
#1377 

*Description of changes:*
* This change refactors the middleware stack class into a factory. So that the private data are stored in closure instead of class "private" members. Now, you can instantiate a middleware stack by `constructStack<Input, Output>();` instead of `new MiddlewareStack<Input, Output>()`.

* Remove `step` from the `addRelativeTo` options. This way the position of the middleware is solely depend on the `toMiddleware` and `relation`(before or after).

* Refactor the middleware stack resolution implementation. Now the resolution algorithm is easier to follow & maintain.

/cc @trivikr 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: packages/middleware-stack/src/MiddlewareStack.ts [hunk 1]
```diff
@@ -1,381 +1,228 @@
...
-
-  use(plugin: Pluggable<Input, Output>) {
-    plugin.applyToStack(this);
-  }
+  };
+
+  const expandRelativeMiddlewareList = (
+    from: Normalized<MiddlewareEntry<Input, Output>, Input, Output>
+  ): MiddlewareEntry<Input, Output>[] => {
+    const expandedMiddleareList: MiddlewareEntry<Input, Output>[] = [];
```
Comment (trivikr) [inline packages/middleware-stack/src/MiddlewareStack.ts:93]:
```suggestion
    const expandedMiddlewareList: MiddlewareEntry<Input, Output>[] = [];
```

File: packages/types/src/middleware.ts [hunk 2]
```diff
@@ -360,10 +314,12 @@ export interface MiddlewareStack<Input extends object, Output extends object> {
...
 
   /**
-   * Create a shallow clone of this list. Step bindings and handler priorities
+   * Create a shallow clone of this stack. Step bindings and handler priorities
    * and tags are preserved in the copy.
+   * Users can optionally supply a base stack that will be merged into the clone
+   * output but its middleware will take preference.
    */
-  clone(): MiddlewareStack<Input, Output>;
+  clone(base?: MiddlewareStack<Input, Output>): MiddlewareStack<Input, Output>;
```
Comment (trivikr) [inline packages/types/src/middleware.ts:322]: Is there a requirement for merging stacks?

If yes, this should a separate operation, and not clone.

File: packages/middleware-stack/src/MiddlewareStack.ts [hunk 3]
```diff
@@ -1,381 +1,228 @@
...
-          resolvedPrev.next = resolvedCurr;
-          if (resolvedCurr.next) resolvedCurr.next.prev = resolvedCurr;
-          resolvedCurr.prev = resolvedPrev;
+  const getMiddlewareList = (): Array<MiddlewareType<Input, Output>> => {
+    const normalizedAbsoluteEntries: Normalized<AbsoluteMiddlewareEntry<Input, Output>, Input, Output>[] = [];
+    const normalizedRelativeEntries: Normalized<RelativeMiddlewareEntry<Input, Output>, Input, Output>[] = [];
+    const normalizedEntriesNameMap: {
+      [middlewareName: string]: Normalized<MiddlewareEntry<Input, Output>, Input, Output>;
+    } = {};
+    absoluteEntries.forEach((entry) => {
```
Comment (trivikr) [inline packages/middleware-stack/src/MiddlewareStack.ts:121]: newline for better readability
```suggestion

    absoluteEntries.forEach((entry) => {
```

Ditto on lines 130, 139, 157 and other places.

File: packages/middleware-stack/src/MiddlewareStack.ts [hunk 4]
```diff
@@ -1,381 +1,228 @@
...
+      };
+      if (name) {
+        if (entriesNameSet.has(name)) {
+          throw new Error(`Duplicated middleware name '${name}'`);
+        }
+        entriesNameSet.add(name);
+      }
+      absoluteEntries.push(entry);
+    },
+    addRelativeTo: (middleware: MiddlewareType<Input, Output>, options: HandlerOptions & RelativeLocation) => {
```
Comment (trivikr) [inline packages/middleware-stack/src/MiddlewareStack.ts:185]: newline for better readability
```suggestion

    addRelativeTo: (middleware: MiddlewareType<Input, Output>, options: HandlerOptions & RelativeLocation) => {
```

Also on lines 203, 204, 207 and more

File: packages/middleware-stack/src/MiddlewareStack.ts [hunk 5]
```diff
@@ -1,381 +1,228 @@
...
+      const entry: AbsoluteMiddlewareEntry<Input, Output> = {
+        name,
+        step,
+        tags,
+        priority,
+        middleware,
+      };
+      if (name) {
+        if (entriesNameSet.has(name)) {
+          throw new Error(`Duplicated middleware name '${name}'`);
```
Comment (trivikr) [inline packages/middleware-stack/src/MiddlewareStack.ts:179]:
```suggestion
          throw new Error(`Duplicate middleware name '${name}'`);
```

File: packages/middleware-stack/src/MiddlewareStack.ts [hunk 6]
```diff
@@ -1,381 +1,228 @@
...
+  const stack = {
+    add: (middleware: MiddlewareType<Input, Output>, options: HandlerOptions & AbsoluteLocation = {}) => {
+      const { name, step = "initialize", tags, priority = "normal" } = options;
+      const entry: AbsoluteMiddlewareEntry<Input, Output> = {
+        name,
+        step,
+        tags,
+        priority,
+        middleware,
+      };
```
Comment (trivikr) [inline packages/middleware-stack/src/MiddlewareStack.ts:176]: nit: some redundant variables can be reduced.
```suggestion
      const { name } = options;
      const entry: AbsoluteMiddlewareEntry<Input, Output> = {
        step: "initialize",
        priority: "normal",
        middleware,
        ...options,
      };
```

Ditto in addRelativeTo

File: packages/middleware-stack/src/MiddlewareStack.spec.ts [hunk 7]
```diff
@@ -1,277 +1,327 @@
...
+function getConcatMiddleware(message: string) {
+  return (next: FinalizeHandler<input, output>): InitializeHandler<input, output> => {
+    return (args: any) =>
       next({
         ...args,
         input: args.input.concat(message),
+        request: undefined as any,
       });
   };
 }
```
[the remark this answers is not in the document]
Comment (AllanFly120) [reply packages/middleware-stack/src/MiddlewareStack.spec.ts:25]: Yup, I should pay more attention to it

File: packages/types/src/middleware.ts (hunk 2 above)
Comment (AllanFly120) [reply packages/types/src/middleware.ts]: Addressed in 1e15a4b

**Pull request merged**